Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+        case let .int32(i):
+            return Decimal128(String(i))
+        case let .double(d):
+            return Decimal128(String(d))
+        default:
+            return nil
+        }
+    }
+
+    /// Return this BSON as an `[T]` if possible, transforming each element via the provided closure.
+    /// If the closure returns nil, the value will be discarded from the resultant array.
+    public func asArray<T>(_ transform: (BSON) -> T?) -> [T]? {
+        return self.arrayValue?.compactMap(transform)
+    }
+}
+
+/// Extension providing the internal API of `BSON`
+extension BSON {
+    /// List of all BSONValue types. Can be used to exhaustively check each one at runtime.
+    internal static var allBSONTypes: [BSONValue.Type] = [
+        BSONNull.self,
+        BSONUndefined.self,
+        MinKey.self,
+        MaxKey.self,
+        Symbol.self,
+        Double.self,
+        String.self,
+        Document.self,
+        Binary.self,
+        ObjectId.self,
+        Bool.self,
+        Date.self,
+        RegularExpression.self,
+        DBPointer.self,
+        Code.self,
+        CodeWithScope.self,
+        Int32.self,
+        Timestamp.self,
+        Int64.self,
+        [BSON].self,
+        Decimal128.self
+    ]
+
+    /// Get the associated `BSONValue` to this `BSON` case.
+    internal var bsonValue: BSONValue {
+        switch self {
+        case .null:
+            return BSONNull()
+        case .undefined:
+            return BSONUndefined()
+        case .minKey:
+            return MinKey()
+        case .maxKey:
+            return MaxKey()
+        case let .symbol(v):
+            return v
+        case let .double(v):
+            return v
+        case let .string(v):
+            return v
+        case let .document(v):
+            return v
+        case let .binary(v):
+            return v
+        case let .objectId(v):
+            return v
+        case let .bool(v):
+            return v
+        case let .date(v):
+            return v
+        case let .regex(v):
+            return v
+        case let .dbPointer(v):
+            return v
+        case let .code(v):
+            return v
+        case let .codeWithScope(v):
+            return v
+        case let .int32(v):
+            return v
+        case let .timestamp(v):
+            return v
+        case let .int64(v):
+            return v
+        case let .array(v):
+            return v
+        case let .decimal128(v):
+            return v
+        }
+    }
+}
+
+extension BSON: ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self = .string(value)
+    }
+}
+
+extension BSON: ExpressibleByBooleanLiteral {
+    public init(booleanLiteral value: Bool) {
+        self = .bool(value)
+    }
+}
+
+extension BSON: ExpressibleByFloatLiteral {
+    public init(floatLiteral value: Double) {
+        self = .double(value)
+    }
+}
+
+extension BSON: ExpressibleByIntegerLiteral {
```

Comment:
so if I have an `Int` saved in a variable and I want to turn it into a BSON do I have to cast it?

e.g. like
```swift
let x = 1
let bson = BSON.int64(Int64(x))
```

Reply:
Good catch. In the tests I was using `BSON(integerLiteral: x)` but meant to add a better initializer for this (e.g. `BSON(x)` or `BSON(from int: Int)`. For now I've just gone with the no-label option.